geopandas/tests/test_geodataframe.py

```diff
@@ -503,3 +504,134 @@ def test_pickle(self):
+        data = {"A": pd.Series(range(3)), "B": np.arange(3.0),
+                "geometry": GeoSeries([Point(x, x) for x in range(3)])}
+        with pytest.raises(ValueError):
+            GeoDataFrame(data, index=[1, 2])
+
+    def test_dict_specified_geometry(self):
+
+        data = {"A": range(3), "B": np.arange(3.0),
+                "other_geom": [Point(x, x) for x in range(3)]}
+
+        df = GeoDataFrame(data, geometry='other_geom')
+        check_geodataframe(df, 'other_geom')
+
+        with pytest.raises(ValueError):
+            df = GeoDataFrame(data, geometry='geometry')
+
+        # when no geometry specified -> works but raises error once
+        # trying to access geometry
+        df = GeoDataFrame(data)
+
+        with pytest.raises(AttributeError):
+            df.geometry
+
+        df = df.set_geometry('other_geom')
+        check_geodataframe(df, 'other_geom')
+
+        # combined with custom args
+        df = GeoDataFrame(data, geometry='other_geom',
+                          columns=['B', 'other_geom'])
+        check_geodataframe(df, 'other_geom')
+        assert_index_equal(df.columns, pd.Index(['B', 'other_geom']))
+        assert_series_equal(df['B'], pd.Series(np.arange(3.), name='B'))
+
+        df = GeoDataFrame(data, geometry='other_geom',
+                          columns=['other_geom', 'A'])
+        check_geodataframe(df, 'other_geom')
+        assert_index_equal(df.columns, pd.Index(['other_geom', 'A']))
+        assert_series_equal(df['A'], pd.Series(range(3), name='A'))
+
+    def test_array(self):
+        data = {"A": range(3), "B": np.arange(3.0),
+                "geometry": [Point(x, x) for x in range(3)]}
+        a = np.array([data['A'], data['B'], data['geometry']], dtype=object).T
+
+        df = GeoDataFrame(a, columns=['A', 'B', 'geometry'])
+        check_geodataframe(df)
+
+        df = GeoDataFrame(a, columns=['A', 'B', 'other_geom'],
+                          geometry='other_geom')
+        check_geodataframe(df, 'other_geom')
+
+    def test_from_frame(self):
+        data = {"A": range(3), "B": np.arange(3.0),
+                "geometry": [Point(x, x) for x in range(3)]}
+        gpdf = GeoDataFrame(data)
+        pddf = pd.DataFrame(data)
+        check_geodataframe(gpdf)
+        assert type(pddf) == pd.DataFrame
+
+        for df in [gpdf, pddf]:
+
+            res = GeoDataFrame(df)
+            check_geodataframe(res)
+
+            res = GeoDataFrame(df, index=pd.Index([0, 2]))
+            check_geodataframe(res)
+            assert_index_equal(res.index, pd.Index([0, 2]))
+            assert res['A'].tolist() == [0, 2]
+
+            res = GeoDataFrame(df, columns=['geometry', 'B'])
+            check_geodataframe(res)
+            assert_index_equal(res.columns, pd.Index(['geometry', 'B']))
+
+            with pytest.raises(ValueError):
+                GeoDataFrame(df, geometry='other_geom')
+
+        data = {"A": range(3), "B": np.arange(3.0),
```

Comment:
Any reason not to split this into a second test, analogous to above with `test_dict` and `test_dict_specified_geometry`?